agent/engine/docker_image_manager.go

```diff
@@ -303,14 +309,147 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
 	imageManager.updateLock.Lock()
 	defer imageManager.updateLock.Unlock()
 
+	var numECSImagesDeleted int
 	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates())
 	for i := 0; i < imageManager.numImagesToDelete; i++ {
 		err := imageManager.removeLeastRecentlyUsedImage(ctx)
+		numECSImagesDeleted = i + 1
 		if err != nil {
 			seelog.Infof("End of eligible images for deletion: %v; Still have %d image states being managed", err, len(imageManager.getAllImageStates()))
 			break
 		}
 	}
+	if imageManager.deleteNonECSImagesEnable {
+		// remove nonecs containers
+		imageManager.removeNonECSContainers(ctx)
+		// remove nonecs images
+		var nonECSImagesNumToDelete = imageManager.numImagesToDelete - numECSImagesDeleted
+		imageManager.removeNonECSImages(ctx, nonECSImagesNumToDelete)
+	}
+}
+
+func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Context) {
+	nonECSContainersIDs, err := imageManager.nonECSContainersIDs(ctx)
+	if err != nil {
+		fmt.Errorf("Cannoe get nonECSContainersIDs Error")
+	}
+	var nonECSContainerRemoveAvailableIDs []string
+	for _, id := range nonECSContainersIDs {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if response.State.Status == "exited" && time.Now().Sub(response.State.FinishedAt) > imageManager.nonECSContainerCleanupWaitDuration {
+			nonECSContainerRemoveAvailableIDs = append(nonECSContainerRemoveAvailableIDs, id)
+		}
+	}
+	var numNonECSContainerDeleted = 0
+	for _, id := range nonECSContainerRemoveAvailableIDs {
+		if numNonECSContainerDeleted == imageManager.numNonECSContainersToDelete {
+			break
+		}
+		err := imageManager.client.RemoveContainer(ctx, id, dockerclient.RemoveContainerTimeout)
+		if err == nil {
+			numNonECSContainerDeleted++
+		} else {
+			seelog.Info("nonECS Container cannot be removed %v", id)
+			continue
+		}
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSContainersIDs(ctx context.Context) ([]string, error) {
+	ECSContainersIDs := imageManager.state.GetAllContainerIDs()
+	var allContainersIDs []string
+	listContainersResponse := imageManager.client.ListContainers(ctx, true, dockerclient.ListContainersTimeout)
```

Comment:
i think ListContainers needs to be before GetAllContainerIDs. otherwise if a new ECS container is started between GetAllContainerIDs and ListContainers, it is treated as a non-ECS container